### ISSUE TITLE
Explicit tests are automatically run in Visual Studio 2019

### ISSUE THREAD

**Issue by Stolpe** [nunit/nunit3-vs-adapter#658]
Hi,

Since switching to Visual Studio 2019, tests marked as `[Explicit]` are run automatically when running all tests, such as when running all tests after build or manually telling Visual Studio to run all tests via the Test Explorer. This worked well on Visual Studio 2017, meaning that explicit tests were not run.

See the repo linked below for a minimal example.

Marking tests with `[Ignore]` still excludes the test from being run (although I do not seem to be able to run those tests even manually, but that's another story)

* NUnit 3.12.0 and NUnit3TestAdapter 3.15.1
* Visual Studio 2019 Pro (16.2.5) and Visual Studio 2017 Enterprise (15.9.16)
* https://github.com/Stolpe/NUnitExplicitTest
* netcoreapp 2.1

**Comment by ebolakid** [comment]
I can confirm that this is also the case on my machine when running the linked repo when running VS 2019. Unfortunately (?) I have the exact samt version (Pro 16.2.5).

**Comment by Stolpe** [comment]
Since there was just some new releases, I updated to the following.
* Visual Studio 2019 Pro 16.3.0
* Microsoft.NET.Test.Sdk 16.3.0
* netcoreapp3.0
The issue still remains, though.

**Comment by shen747** [comment]
I am having the same issue with the VS 16.3.5

Any resolution would be highly appreciated

**Comment by OsirisTerje** [comment]
Short update:  I can confirm there is a change in how Visual Studio calls the adapter that causes this behavior.  We have contacted the Product Group about it, and will come back with how to fix this.

**Comment by ebolakid** [comment]
Thank you for the update, looking forward to a fix in the near future

**Comment by OsirisTerje** [comment]
This issue is caused by a change in behavior in Visual Studio 2019, in how the test execution methods are being called.

The adapter have two methods that the testhost can be called, one "by sources" and one "by testcases".   The first one here was originally meant to be called from 1) command line 2) Visual Studio when running All Tests.  The second one was meant to be called only from Visual Studio when something less than All Tests were selected.

Right now it seems that the second method is being called in all cases.  I have so far not found that the 1st method is being called at all.  I have tested from Visual Studo, from vstest.console and from dotnet test.

The issue here is that there is no way to deduce whether all tests are being run or not.  There is no explicit information on what triggered the execution of the tests.

This means that all tests runs, regardless of how they are started, are perceived as a set of selected tests by the adapter, which means that they fulfill the demand set by the Explicit attribute. 

The Product Group have been informed, and we're waiting for further information from them on how this can be resolved.

**Comment by condron** [comment]
Is there a work-around?

**Comment by OsirisTerje** [comment]
@condron Currently the only way is to use the fact that the Explicit also is a category.  So in a build you can exclude the explicit tests by including an exclusion for any test with this attribute.    From Visual Studio you can use the search/filter in the Test explorer to include or exclude that category.

**Comment by patearl** [comment]
I wonder if this could be approached from the other direction of somehow allowing "-trait:explicit" to be set by default. Perhaps the test engine could query the adapter for a default filter. As another approach, a mechanism to provide a default the test search on a team-wide basis could work. Perhaps they could even directly set the default on the search to a "known by convention" trait that could be provided by the various test adapters. Of course another approach is to make "explicitness" a first class citizen of the UI, but I doubt they would be willing to do that since some test frameworks have no such concept.

**Comment by sodablue** [comment]
I came up with a workaround/hack.   I found the Ignored attribute works which means the runner can skip tests.   I created my own DebugExplicitAttribute from the ExplicitAttribute which sets the runstate to ignored if the debugger isn't attached.   This allows me to run the test manually, but running all tests it will be skipped.

see this gist
https://gist.github.com/sodablue/6941abcd6855e71b4393118c3e01bf2f

**Comment by OsirisTerje** [comment]
@sodablue  Very nicely done!   Would you consider a blog post for this ?   I will create more visibility to this, and I'll make sure to reference it here somewhere.

**Comment by sodablue** [comment]
I'd have to have a blog. :-)   I guess I used to have one like 20 years ago, but

I'd encourage you to write it up.   I just wanted to share it here as a possible workaround, and thought maybe the fact Ignore works might help someone troubleshoot the issue with Visual Studio.

**Comment by OsirisTerje** [comment]
Ok, I'll do it, will anyway credit you.   

For this issue, the thought is to change it into handling one by one.  In earlier VS you could take a higher level, and it would run all explicit tests.  That is what not works now, as everything comes in that way.  What we can do is to allow explicits to run if they are  1) selected only one test that is explicits  2) all tests selected are Explicits.    Until that your workaround is a very good one, I also like it as it shows one more way to extend NUnit :-)

**Comment by olivier-spinelli** [comment]
Another funny workaround: When you really want a Test to run, press the "Ctrl" key... This is really stupid but it works well. Tests must be written this way (note that the [Explicit] attribute is no more required but makes the intent more... explicit):

```csharp
        [Explicit]
        [Test]
        public void this_is_an_explicit_test()
        {
            Assume.That( TestHelper.IsExplicitAllowed, "Press Ctrl key to allow this test to run." );
            /// ... code come here.
        }
```

**Comment by arnileibovits** [comment]
This is still not fixed in the latest version 16.5.1.

The issue was opened on Sep 17, 2019...

**Comment by aolszowka** [comment]
Ok I have taken a stab at understanding the code base, spun up a new fresh VM with BOTH Visual Studio 2017 AND Visual Studio 2019, and want to share my findings.

Lets start with the comment by @OsirisTerje 

>This issue is caused by a change in behavior in Visual Studio 2019, in how the test execution methods are being called.
>
>The adapter have two methods that the testhost can call, one "by sources" and one "by testcases". The first one here was originally meant to be called from 1) command line 2) Visual Studio when running All Tests. The second one was meant to be called only from Visual Studio when something less than All Tests were selected.
>
>Right now it seems that the second method is being called in all cases. I have so far not found that the 1st method is being called at all.

The two methods are:

1. I will call this the "IEnumerable<string>" or "string" version https://github.com/nunit/nunit3-vs-adapter/blob/152b1c1e430414495d0ad2d4ed22f59063a8f531/src/NUnitTestAdapter/NUnit3TestExecutor.cs#L87
2. I will call this the "IEnumerable<TestCase>" or "TestCase" version https://github.com/nunit/nunit3-vs-adapter/blob/152b1c1e430414495d0ad2d4ed22f59063a8f531/src/NUnitTestAdapter/NUnit3TestExecutor.cs#L130

In Visual Studio 2017 Enterprise 15.9.23 opening up the above test project (created by @Stolpe but converted to Full .NET Framework because 2017 does not have a .NET Core 2.1 SDK) and clicking "Run All":

![image](https://user-images.githubusercontent.com/2592881/82276054-739fd500-9941-11ea-89c6-c31faeced566.png)

Will invoke `RunTests(IEnumerable<string>, IRunContext , IFrameworkHandle)` in this version of the code path the Adapter assumes that it has been passed a `Settings.Where` that will be used to filter the tests:

https://github.com/nunit/nunit3-vs-adapter/blob/152b1c1e430414495d0ad2d4ed22f59063a8f531/src/NUnitTestAdapter/NUnit3TestExecutor.cs#L109

This is done (it is assumed) to allow further filtering, however in practice from Visual Studio 2017 when this is ran, via clicking `Run All` the value of `Settings.Where` is null. The Test Adapter takes this to mean that you want to pass no filter to the underlying NUnit Runner:

https://github.com/nunit/nunit3-vs-adapter/blob/152b1c1e430414495d0ad2d4ed22f59063a8f531/src/NUnitTestAdapter/NUnitTestFilterBuilder.cs#L53-L54

In other words you want to invoke the default behavior of the NUnit Runner which is to:

* For Explicit (https://github.com/nunit/docs/wiki/Explicit-Attribute) "causes a test or test fixture to be skipped unless it is explicitly selected for running. The test or fixture will be run if it is selected by name or if it is included by use of a filter."
* For Ignore (https://github.com/nunit/docs/wiki/Ignore-Attribute) " Ignored tests are displayed by the runners as warnings in order to provide a reminder that the test needs to be corrected or otherwise changed and re-instated."

Which in our case is what is desired.

**tl;dr;** This worked by relying on implementation behavior of the `IEnumerable<string>` version of this call to determine if tests were explicitly selected or not.

However in Visual Studio 2019 (at least as of Enterprise 16.5.5) clicking Run All (Boxed in Red):

![image](https://user-images.githubusercontent.com/2592881/82276807-8915fe80-9943-11ea-8511-ffee868e87c2.png)

Invokes the `IEnumerable<TestCase>` version of `RunTests` in this version as mentioned above it is assumed that:

>...was meant to be called only from Visual Studio when something less than All Tests were selected.

Because the Test Adapter:

>[has] no way to deduce whether all tests are being run or not. There is no explicit information on what triggered the execution of the tests.

The Test Adapter treats the test:

>as a set of selected tests

Which, in other words means that the test adapter assumes you asked for these explicit tests to be ran.

### What can be done to fix this?

1. (Moonshot) We file a Developer Community Post to note the change in behavior for Test Adapters Between VS 2017 and 2019 and hope they revert back to the 2017 behavior.
   - This unfortunately is unlikely to get any real traction, this probably broke early in the Visual Studio 2019 life cycle, but was never narrowed down / reported early enough for them to warrant a fix. Even if they did fix it it would probably break someone else who was depending on this behavior.
   - That being said a close reading of the specification (https://github.com/microsoft/vstest-docs/blob/master/RFCs/0004-Adapter-Extensibility.md) shows that they broke it in VS 2019.
   - I will still make a run at this and file an issue then update this post with a link.

2. (Moonshot) We file a Developer Community Post to ask for a feature enhancement to the Adapter Extensibility specification to give us some type of context as to how the test is being run (perhaps off of `IRunContext`?)
   - Again this might be too steep an ask from Microsoft at this point. The thing that makes this difficult is that MSTest does not have this concept of `Explicit`
   - However with this we could fix the behavior to be much more explicit

3. (Realistic) We could change the behavior of the Test Adapter when we receive multiple tests to the `IEnumerable<TestCase>` overload of `RunTests` in any number of ways
   - If Multiple Tests are Received, Filter out any Tests with the `Explicit`  `Microsoft.VisualStudio.TestPlatform.ObjectModel.Trait` within the framework (it is unclear if this is what @patearl had in mind or something more).
     - This will break anyone who was relying on the behavior of selecting multiple tests, we can lessen the burden by improving the heuristics in various ways: for example we could say that if multiple tests were selected and they all contained the `Explicit` Trait ignore our filtering logic.
   - We could apply some other advanced heuristic; such as: looking at the number of tests being asked to execute, comparing it to the total number of tests in the assembly and if they are equal then apply the filtering to drop `Explicit` tests.
     - It is unclear to me at this juncture what the cost of this would be, however I suspect it would be low because in theory we already have had to "Discover" the contents of the test in question, however if the `IEnumerable<TestCase>` passed to us is lazily evaluated performing a `.Count()` might be very expensive.
   - Regardless of the option chosen we probably need yet another setting to turn off (opt out) of these changes, which is never free to do.

I am willing to provide a pull request that implements the simple heuristic of dropping any tests that contain the `Explicit` trait if there are more than one test being invoked and any other test does not contain the `Explicit` trait, along with the option to opt out of the behavior. However before doing so I'd want some type of assurance that the likely-hood of accepting the functional change would be high before investing more time.

At very least I am willing to commit to opening a Developer Community post for bullet point 1; if there is interest I would draft one up for bullet point 2.

Let me know.

**Comment by OsirisTerje** [comment]
@aolszowka  Thanks a lot for this, awesome!  

1.  I doubt MS will go back. And we did use a side effect to handle explicit testing.  I have been looking and preparing for changing the algorithm for Explicit testing.  In the code for 3.17 the XML is moved into its own "adapter to nunit engine", so that we get a cleaner interface, and a structure that are more easily traversed.  The idea is then to see if all tests are explicit , if so, then it is an explicit test run, and those are executed.  If not, then all explicit tests are removed from the pre-execution-discovery, and thus not executed.  This is easy to do for single tests, but since a testfixture can be attributed, affecting all children, we need to traverse the hierarchy.  That can be costly in processing time if we don't have a good structure.      So, I think this is the way we have to go here.

2. I have asked for this earlier, with no positive response, so awesome if you create a dev com issue on that.  I'll for sure vote it up :-)  (We have also up to now been sort of guessing on WHO calls us......, would be nice to have that explicitly stated)

3. This is kind of the same thing I have commented on above, so yes, I think this is the sort of way we need to go.

If you would like to participate implementing this, that would be awesome.   Take a look at the branch 'refactorengineadapter'.   The next thing up there is the NUnitResult class, which currently works as a temporary anti-corruption layer.   Inside the testConverter is being called, which converts the NUnit XML into MS TestObjects.  They need to go into some other NunitTestCase-type of object, which can represent the whole hierarchy.   The idea is to create a structure for this on the side of the existing (so that one can opt in/out), and use that to do the checks for Explicit or not.  The same structure would also make it much more easy to implement other features where we rely on the structure.  There are some nasty code in the adapter today which works on that.

**Comment by aolszowka** [comment]
Just to Update:

1. DevCom Issue Posted Here: https://developercommunity.visualstudio.com/content/problem/1038508/visual-studio-2019-test-explorer-no-longer-invokes.html

2. DevCom Idea Posted Here: https://developercommunity.visualstudio.com/idea/1038512/test-adapters-should-have-information-about-how-te.html

I will loop back around to take a look at the 'refactorengineadapter' branch after I have exhausted issue tracking a few others that burn us pretty bad. Thank you for being patient with me, I am getting used to the code base and how to interact with GitHub and this project :(

**Comment by plioi** [comment]
FWIW I never hear back from the developcommunity forum, but I do occasionally hear back when I post issues to https://github.com/microsoft/vstest/issues

**Comment by OsirisTerje** [comment]
Resolved with https://github.com/nunit/nunit3-vs-adapter/pull/763

**Comment by lbmaian** [comment]
@OsirisTerje I've just updated to version 3.17 (via online update within VS), and VS 2019 is still automatically running explicit tests. The release notes for version 3.17 also neither link this issue nor mention "explicit" at all. Are you sure this issue is fixed?

If it matters, I'm working with a multi-TFM project.

**Comment by OsirisTerje** [comment]
@lbmaian You're correct it is not part of 3.17.   There was so many other issues in 3.17, and the changes for this issue requires a large refactoring.  So it was pushed to 4.0.  That said, an alpha-version of the 4.0 will be pushed out to nuget real soon now (meaning a day or something, if all goes well).  You can grab a dev version at https://www.myget.org/feed/nunit/package/nuget/NUnit3TestAdapter  (4.0.0-dev-01491).   What is known there is some trouble with data driven tests.  I'm trying to weed that out before a public alpha goes out. 

PS:  If you look at the Milestone for this issue, you will see it is 4.0, not 3.17 :-)   )

**Comment by OsirisTerje** [comment]
@Stolpe @lbmaian @aolszowka @plioi @arnileibovits @olivier-spinelli Alpha.1 released now: https://www.nuget.org/packages/NUnit3TestAdapter/4.0.0-alpha.1

**Comment by aolszowka** [comment]
I'm getting lots of NRE's:

```
Test 0-1136 not found in cache
Test 0-1136 not found in cache
Test 0-1136 not found in cache
Error processing Validate("5","5","00000",AreaCodeMustBeThreeDigits, ExchangeCodeMustBeThreeDigits, SubscriberNumberMustBeFourDigits) event for REDACTED.Tests.Core.Utilities.Telephony.USPhoneNumberValidatorTests.Validate("5","5","00000",AreaCodeMustBeThreeDigits, ExchangeCodeMustBeThreeDigits, SubscriberNumberMustBeFourDigits)
System.NullReferenceException: Object reference not set to an instance of an object.
   at NUnit.VisualStudio.TestAdapter.NUnitEventListener.TestFinished(INUnitTestEventTestCase resultNode) in D:\repos\NUnit\nunit3-vs-adapter\src\NUnitTestAdapter\NUnitEventListener.cs:line 169
   at NUnit.VisualStudio.TestAdapter.NUnitEventListener.OnTestEvent(String report) in D:\repos\NUnit\nunit3-vs-adapter\src\NUnitTestAdapter\NUnitEventListener.cs:line 92
NUnit Adapter 4.0.0.0: Test execution complete
```

Nothing appears to run?

**Comment by OsirisTerje** [comment]
I'm running the repro here:  https://github.com/nunit/nunit3-vs-adapter.issues/tree/master/Issue658/Explicit  (Taken from the one given by the reporter, with a few extras)

And I do get this:
![image](https://user-images.githubusercontent.com/203432/87255972-80e3c680-c48f-11ea-99ed-d41fe21d8eac.png)

Are you sure you have a real rebuild here?
Is it the same repro?

**Comment by aolszowka** [comment]
@OsirisTerje Sorry no, this is in our production code base, I just cloned the test you linked above and that worked as expected.

Means some regression elsewhere... I'll check into it first thing Monday Morning. Anywhere in particular I should be looking first?

Thank you for jumping on this and getting it done.

**Comment by lbmaian** [comment]
@OsirisTerje Ah thanks for the clarification. I was confused by the timing of the closing of this issue, the merging of the linked PR, and the 3.17 release.

**Comment by OsirisTerje** [comment]
@aolszowka take a look in the execution dump file.  It has more info than earlier.   You see also what comes from vs. Also, you should be sble to find the ids thete

**Comment by aolszowka** [comment]
@OsirisTerje I just got into the office and had fired this up, I have it repo'ing in a smaller set, can you point to the execution dump file documentation again? I feel like I needed to set an environment variable or something?

**Comment by OsirisTerje** [comment]
Go to the tips and tricks section in the docs for the adapter.  Set it in the runsettings file.

**Comment by aolszowka** [comment]
@OsirisTerje Ugh I have narrowed it down (after a bunch of run around); its because our Production Code base is still referencing NUnit 3.11; it reproduces with any test. Upgrading to NUnit 3.12 makes this disappear.

To keep this issue clean I will open up a new issue and  will submit a PR that has the repo to your 
nunit3-vs-adapter.issues repo.

I apologize to all on this thread getting spammed by this.

**Comment by OsirisTerje** [comment]
@aolszowka But this is awesome!   What you say, is that 3.17 doesn't work with NUnit earlier than 3.12 :-)  We need to get that explicitly stated somewhere.  

Really appreciate you checking this out, and reporting back on it!

**Comment by aolszowka** [comment]
@OsirisTerje Opened #768 sorry about that :( I'll start the ball internally trying to get ourselves updated.

**Comment by OsirisTerje** [comment]
Thanks again @aolszowka .   We'll check that out and get it confirmed there, and how best to go about it.

**Comment by Stolpe** [comment]
Hi. I'm back from paternity leave and I find that this issue if Closed as fixed. When I try my minimal test project, the issue indeed seems to be fixed. Great work!
However, when I update to 4.0.0-alpha.1 of the test adapter in some of our projects here at work, I find that the issue still remains for tests which uses `[TestCaseSource]`. I added a branch in my original branch at https://github.com/Stolpe/NUnitExplicitTest/tree/testcasesource. I also verified that this works as intended in Visual Studio 2017 (using the latest alpha). 
@OsirisTerje, perhaps you should consider reopening this issue.
